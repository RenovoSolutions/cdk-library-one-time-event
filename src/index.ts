import * as events from '@aws-cdk/aws-events';
import * as cdk from '@aws-cdk/core';

export interface IOnDeployProps {
  /**
   * The number of minutes to add to the current time when generating the expression.
   * Should exceed the expected time for the appropriate resources to converge.
   * Defaults to 10 mins.
   */
  offsetMinutes?: number;
}

export interface IAtProps {
  /**
   * The future date to use for one time event.
   */
  date: Date;
}

export function dateToCron(date:Date) {
  const minutes = date.getUTCMinutes();
  const hours = date.getUTCHours();
  const days = date.getUTCDay();
  const months = date.getUTCMonth() + 1;
  const years = date.getUTCFullYear();

  return `${minutes} ${hours} ${days} ${months} ? ${years}`;
};

export class OnDeploy extends cdk.Construct {

  public readonly schedule: events.Schedule;

  constructor(scope: cdk.Construct, id: string, props: IOnDeployProps) {
    super(scope, id);

    const date = new Date();
    const minutesToAdd = props.offsetMinutes || 10;
    const future = new Date(date.getTime() + minutesToAdd * 60000);

    this.schedule = events.Schedule.expression('cron(' + dateToCron(future) + ')');
  }
}

export class At extends cdk.Construct {

  public readonly schedule: events.Schedule;

  constructor(scope: cdk.Construct, id: string, props: IAtProps) {
    super(scope, id);

    this.schedule = events.Schedule.expression('cron(' + dateToCron(props.date) + ')');
  }
}
