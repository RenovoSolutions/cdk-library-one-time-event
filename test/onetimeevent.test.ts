import { SynthUtils } from '@aws-cdk/assert';
import { aws_events as events, App, Stack } from 'aws-cdk-lib';
import * as oneTimeEvents from '../src/index';

jest.setSystemTime(new Date('2020-01-01').getTime());

test('Snapshot', () => {
  const app = new App();
  const stack = new Stack(app, 'TestStack', {});

  new events.Rule(stack, 'triggerImmediate', {
    schedule: new oneTimeEvents.OnDeploy(stack, 'schedule', {
      offsetMinutes: 10,
    }).schedule,
  });

  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});
