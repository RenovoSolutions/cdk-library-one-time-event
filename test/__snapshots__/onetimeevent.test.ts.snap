// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "triggerImmediate338B3A22": {
      "Properties": {
        "ScheduleExpression": "cron(33 15 14 1 ? 2020)",
        "State": "ENABLED",
      },
      "Type": "AWS::Events::Rule",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
